src/cheri_pte.sail

```diff
@@ -49,6 +88,62 @@ union PTE_Check = {
   PTE_Check_Failure : (ext_ptw, ext_ptw_fail)
 }
 
+/*
+ * Assuming we're allowed to load from this page, modulate our cap response
+ */
+val checkPTEPermission_LC : (PTE_Bits, Ext_PTE_Bits, ext_ptw) -> PTE_Check effect { escape, rreg }
+function checkPTEPermission_LC(p, e, ext_ptw) = {
+  let res : PTE_Check =
+    if e.CapRead() == 0b0
+    then match (e.CapRead_Gen(), e.CapRead_Mod()) {
+           (0b1, _)  => PTE_Check_Failure(ext_ptw, EPTWF_CAP_ERR),
+           (_, 0b0)  => PTE_Check_Success(ext_ptw_lc_join(ext_ptw, PTW_LC_CLEAR)) /* Clear tag for "unmodified" no-LC case */,
+           (_, 0b1)  => PTE_Check_Success(ext_ptw_lc_join(ext_ptw, PTW_LC_TRAP))  /* Trap on tag load for "modified" no-LC case */
+        }
+    else if e.CapRead_Mod() == 0b0
+         then { if e.CapRead_Gen() == 0b0
+                then PTE_Check_Success(ext_ptw_lc_join(ext_ptw, PTW_LC_OK)) /* Unmodified LC case: go ahead */
+                else PTE_Check_Failure(ext_ptw, EPTWF_CAP_ERR)
+         }
+         else { /* With CapRead_Mod on with LC, use Gen */
+                let gclg : bits(1) = if p.U() == 0b1 then sccsr.gclgu() else sccsr.gclgs();
+                if e.CapRead_Gen() == gclg
+                then PTE_Check_Success(ext_ptw_lc_join(ext_ptw, PTW_LC_OK))
+                else PTE_Check_Success(ext_ptw_lc_join(ext_ptw, PTW_LC_TRAP))
+         };
```

Comment:
This would be better as a big 3-tuple match with the fields and cases in the same order as your helpful table above. As it stands it's hard to match things up, and a bit messy.